File: volume4/7_2_1_3combinatorialsearching/algorithm_l/algorithm_l.rb
# General Algorithms - Combinatronics
# Lexicographic Combination Generator
# From Knuth's The Art of Computer Programming
# Author: Carlos L. Cuenca
# Date: 10/20/20

# --------------
# Driver Program

ARGF.each do |line|

	# As an example, we go through each n choose t
	for t in 1..array.length

		combinations = Array.new(t + 2)

		# Clear the combination index array
		for j in 0..t - 1

			combinations[j] = j

		end

		# Assign the Sentinels
		combinations[t] = array.length
		combinations[t + 1] = 0

		j = 0

		while j < t do

			# Visit the combination. In this case we print the indices
			combinationString = ""

			for index in 0..t - 1

				combinationString += combinations[index].to_s + " "

			end

			puts combinationString

			# Find j
			j = 0

			while(combinations[j] + 1 == combinations[j + 1]) do

				combinations[j] = j

				j += 1

			end

			combinations[j] += 1

		end

	end

end

File: volume4/7_2_1_3combinatorialsearching/algorithm_l/algorithm_l.cpp
/// General Algorithms - Combinatronics
/// Lexicographic Combination Generator
/// Author: Carlos L. Cuenca
/// Date: 10/13/20

#include<iostream>
#include<string>
#include<vector>
#include<cmath>
#include<climits>
#include<algorithm>

/// -----------------
/// Program Constants

const int SUCCESS = 0;

/// --------------
/// Driver Program

int main(int argc, char* argv[]) {

    /// -----------------
    /// Program Variables

	int number;

	/// -------
	/// Program

	std::cin >> number;

	//  We loop through all t combinations
	for(int t = 1; t <= number; t++) {
		
		std::vector<int> combinations(t + 2);

		for(int j = 0; j < t; j++)
			combinations[j] = j;

		combinations[t] = number;
		combinations[t + 1] = 0;

		int j = 0;

		while(j < t) {

			for(int index = 0; index < t; index++)
				std::cout << combinations[index] << " ";

			std::cout << std::endl;

			j = 0;

			while(combinations[j] + 1 == combinations[j + 1]) {

				combinations[j] = j;

				j++;

			}

			combinations[j]++;

		}

	}

	return SUCCESS;

}
